crates/engine/src/rocksdb_engine/transaction.rs

```diff
@@ -1,18 +1,230 @@
+        /// The target key
+        key: Vec<u8>,
+    },
+    /// Delete range operation, it will remove the database entries in the range [from, to)
+    DeleteRange {
+        /// The table name
+        table: String,
+        /// The `from` key
+        from: Vec<u8>,
+        /// The `to` key
+        to: Vec<u8>,
+    },
+}
+
+impl<'db> RocksTransaction<'db> {
+    /// Creates a new `RocksTransaction`
+    pub(super) fn new(
+        db: Arc<OptimisticTransactionDB>,
+        txn: Transaction<'db, OptimisticTransactionDB>,
+        engine_size: Arc<AtomicU64>,
+    ) -> Self {
+        Self {
+            db,
+            txn: Mutex::new(Some(txn)),
+            engine_size,
+            txn_size: AtomicUsize::new(0),
+        }
+    }
+}
+
+#[allow(clippy::unwrap_used, clippy::unwrap_in_result)] // txn is always `Some`
+impl StorageOps for RocksTransaction<'_> {
+    fn write(&self, op: WriteOperation<'_>, _sync: bool) -> Result<(), EngineError> {
+        match op.into() {
+            WriteOperationOwned::Put { table, key, value } => {
+                let cf = self
+                    .db
+                    .cf_handle(table.as_ref())
+                    .ok_or_else(|| EngineError::TableNotFound(table.clone()))?;
+                self.txn
+                    .lock()
+                    .as_ref()
+                    .unwrap()
+                    .put_cf(&cf, &key, &value)
+                    .map_err(EngineError::from)?;
+                let _ignore = self.txn_size.fetch_add(
+                    RocksEngine::max_write_size(table.len(), key.len(), value.len()),
+                    Ordering::Relaxed,
+                );
+            }
+            WriteOperationOwned::Delete { table, key } => {
+                let cf = self
+                    .db
+                    .cf_handle(table.as_ref())
+                    .ok_or_else(|| EngineError::TableNotFound(table.clone()))?;
+                self.txn
+                    .lock()
+                    .as_ref()
+                    .unwrap()
+                    .delete_cf(&cf, key)
+                    .map_err(EngineError::from)?;
+            }
+            WriteOperationOwned::DeleteRange { table, from, to } => {
+                let cf = self
+                    .db
+                    .cf_handle(table.as_ref())
+                    .ok_or_else(|| EngineError::TableNotFound(table.clone()))?;
+                let mode = IteratorMode::From(&from, Direction::Forward);
+                #[allow(clippy::pattern_type_mismatch)] // can't be fixed
+                let kvs: Vec<_> = self
+                    .txn
+                    .lock()
+                    .as_ref()
+                    .unwrap()
+                    .iterator_cf(&cf, mode)
+                    .take_while(|res| {
+                        res.as_ref()
+                            .is_ok_and(|(key, _)| key.as_ref() < to.as_slice())
+                    })
+                    .collect::<Result<Vec<_>, _>>()?;
```

Comment:
If the kvs are iterated later, why are we collecting it?

Reply:
fixed